gold/nanomips.cc

```diff
@@ -5431,11 +5431,30 @@ Nanomips_expand_insn<size, big_endian>::type(
     {
     case elfcpp::R_NANOMIPS_PC25_S1:
       {
-        Valtype value = psymval->value(relobj, r_addend) - address - 4;
-        if (!this->template has_overflow_signed<26>(value))
-          return TT_NONE;
-        break;
+      Valtype value = psymval->value(relobj, r_addend) - address - 4;
+
+      if (parameters->options().expand_long_balc_bc()) {
+        typedef typename elfcpp::Elf_types<size>::Elf_Swxword Signed_valtype;
+        bool offset_within_range = true;
+        // The offset in backward branch should be less than or equal to 0xffff.
+        // The offset in forward branch should be greather than or equal to
+        // 0xff0000. The range is (-33423362, 33423360).
+        if (!this->template has_overflow_signed<26>(value)) {
+          if (static_cast<Signed_valtype>(value) <= 0xfe01fffe ||
```

Comment:
g++ complains about comparing values with different signedness for this expression. Release builds are done with `-Weror -Wall' and so will fail. We do selectively disable warnings if required due to upstream changes, but we'd like to avoid them for our contributions.